ignite/metrics/mean_average_precision.py

```diff
@@ -0,0 +1,420 @@
+    m = len(y_pred)
+    n = len(y)
+
+    ious = torch.zeros(m, n)
+
+    for g in range(n):
+        Y = y[g].tolist()
+        ya = Y[2] * Y[3]
+        iscrowd = crowd[g]
+        for d in range(m):
+            P = y_pred[d].tolist()
+            pa = P[2] * P[3]
+            ious[m - d - 1, g] = 0
+            w = min(P[2] + P[0], Y[2] + Y[0]) - max(P[0], Y[0])
+            h = min(P[3] + P[1], Y[3] + Y[1]) - max(P[1], Y[1])
+            if w <= 0 or h <= 0:
+                continue
+            i = w * h
+            if iscrowd:
+                u = pa
+            else:
+                u = pa + ya - i
+            ious[m - d - 1, g] = i / u
+
+    return ious
+
+
+class MeanAveragePrecision(Metric):
+    r"""Calculates Mean Average Precision (mAP) for object detection data.
+
+    .. math::
+       \text{MeanAveragePrecision} = {1/11}\sum_{r \in {0.0, ..., 1.0}}\text{AP}_r
+
+    where \text{MeanAveragePrecision}_r represents average precision at recall value :math:`r`,
+    where :math:`r` ranges from 0.0 to 1.0.
+
+    More details can be found in `Everingham et al. 2009`__
+
+    __ https://homepages.inf.ed.ac.uk/ckiw/postscript/ijcv_voc09.pdf
+
+    Remark:
+
+        This implementation is inspired by pycocotools which can be found `here`__
+
+        __ https://github.com/cocodataset/cocoapi/
+
+    Args:
+        object_area_ranges: dictionary with area_name as key and a tuple of the form
+            (lower_area_range, upper_area_range) as value, where both values are floats.
+            It contains all area ranges for which AP needs to be computed.
+        num_detection_max: Number of maximum detections allowed per image. This is to limit crowding and repeated
+            detections of the same object.
+        iou_thresholds: Optional list of float IoU thresholds for which AP is computed (default: [.5:.05:.95]).
+        rec_thresholds: Optional list of float values to which AP is computed for averaging (default: [0:.01:1]).
+        output_transform: a callable that is used to transform the
+            :class:`~ignite.engine.engine.Engine`'s ``process_function``'s output into the
+            form expected by the metric. This can be useful if, for example, you have a multi-output model and
+            you want to compute the metric with respect to one of the outputs.
+            By default, metrics require the output as ``(y_pred, y)`` or ``{'y_pred': y_pred, 'y': y}``.
+        device: specifies which device updates are accumulated on. Setting the
+            metric's device to be the same as your ``update`` arguments ensures the ``update`` method is
+            non-blocking. By default, CPU.
+
+    Example:
+
+        .. code-block:: python
+
+            import torch
+            from ignite.metrics import MeanAveragePrecision
+
+            # Detection Format:
+            # (id, class, xmin, ymin, width, height, area, crowd, ignore/confidence)
```

Comment:
area, crowd, ignore are not so trivial. Could you explain ? This docstring is the root of the doc, it should be extremely clear.

Reply:
From https://github.com/pytorch/vision/blob/759c5b6dc2dfe46735bf724839f815b5cd79af09/torchvision/models/detection/mask_rcnn.py#L27

It would be simpler to handle a dict format rather than a plain tensor format. What do you think ? It would be easier to have optional crowd, ignore, etc.